auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/external/DeltaUserRegistrationController.kt

```diff
@@ -0,0 +1,241 @@
+        val emailAddress = formParameters["emailAddress"].orEmpty().lowercase()
+        val confirmEmailAddress = formParameters["confirmEmailAddress"].orEmpty().lowercase()
+
+        val firstNameErrors = ArrayList<String>()
+        val lastNameErrors = ArrayList<String>()
+        val emailAddressErrors = ArrayList<String>()
+        val confirmEmailAddressErrors = ArrayList<String>()
+
+        if (firstName.isEmpty()) firstNameErrors.add(firstNameEmpty)
+        if (lastName.isEmpty()) lastNameErrors.add(lastNameEmpty)
+
+        // Delta has had issues with XSS in the past, so we disallow these characters as an extra precaution
+        if (hasDisallowedSpecialCharacters(firstName)) firstNameErrors.add(firstNameContainsDisallowedCharacters)
+        if (hasDisallowedSpecialCharacters(lastName)) lastNameErrors.add(lastNameContainsDisallowedCharacters)
+        if (hasDisallowedSpecialCharacters(emailAddress)) emailAddressErrors.add(emailContainsDisallowedCharacters)
+
+        var organisations: List<Organisation> = listOf()
+        if (emailAddress.isEmpty()) emailAddressErrors.add(emailAddressEmpty)
+        else {
+            if (!emailAddressChecker.hasValidEmailFormat(emailAddress)) emailAddressErrors.add(notAnEmailAddress)
+            else {
+                organisations = organisationService.findAllByDomain(emailToDomain(emailAddress))
+                if (!emailAddressChecker.hasKnownNotRetiredDomain(emailAddress, organisations)) emailAddressErrors.add(
+                    notAKnownDomain
+                )
+            }
+        }
+        if (confirmEmailAddress.isEmpty()) confirmEmailAddressErrors.add(confirmEmailAddressEmpty)
+        else if (confirmEmailAddress != emailAddress) confirmEmailAddressErrors.add(notEqualEmails)
+
+        if (hasErrors(firstNameErrors, lastNameErrors, emailAddressErrors, confirmEmailAddressErrors)) {
+            return call.respondRegisterPage(
+                firstName,
+                lastName,
+                emailAddress,
+                confirmEmailAddress,
+                firstNameErrors,
+                lastNameErrors,
+                emailAddressErrors,
+                confirmEmailAddressErrors
+            )
+        } else {
+            val ssoClientMatchingEmailDomain = ssoConfig.ssoClients.firstOrNull {
+                it.required && emailAddress.lowercase().endsWith(it.emailDomain)
+            }
+            if (ssoClientMatchingEmailDomain != null) {
+                return call.respondRedirect(
+                    deltaRouteWithEmail(
+                        deltaConfig.deltaWebsiteUrl,
+                        ssoClientMatchingEmailDomain.internalId,
+                        emailAddress
+                    )
+                )
+            }
+
+            val registration = Registration(firstName, lastName, emailAddress)
+            lateinit var registrationResult: RegistrationService.RegistrationResult
```

Comment:
```suggestion
            val registrationResult: RegistrationService.RegistrationResult
```
Kotlin should be able to figure out it's always assigned exactly once before it's used

Reply:
It doesn't like this because I've changed the error handling slightly and use it there so it thinks it could be not initialized although it can't be